openest/test/generate/test_functions.py

```diff
@@ -67,3 +90,54 @@ def test_spaninstabase():
             np.testing.assert_equal(yearresult[1], 1.)
         if yearresult[0] == 2:
             np.testing.assert_equal(yearresult[1], 1.)
+
+
+class TestClip:
+    """Basic tests for openest.generate.functions.Clip
+    """
+    def test_units_append(self):
+        """Tests that Clip instances correctly append units from the original subcalc
+        """
+        unit = ['fakeunit']
+        subcalc_mock = MockApplication(years=[0], values=[1.0], unitses=unit)
+        clipped_calc = Clip(subcalc_mock, 0.0, 1.0)
+        assert clipped_calc.unitses == unit * 2
+
+    @pytest.mark.parametrize(
+        "in_years,in_values,expected",
+        [
+            ([0], [-0.5], [0, 0.0, -0.5]),
+            ([1], [0.0], [1, 0.0, 0.0]),
+            ([2], [0.5], [2, 0.5, 0.5]),
+        ],
+        ids=['outside clip', 'clip edge', 'inside clip'],
+    )
+    def test_apply(self, in_years, in_values, expected):
+        """Test self.apply() clips values in, out, and on edge of interval
+        """
+        subcalc_mock = MockApplication(
+            years=in_years,
+            values=in_values,
+            unitses=['fakeunit'],
+        )
+        clipped_calc = Clip(subcalc_mock, 0.0, 1.0)
+        victim_gen = clipped_calc.apply('foobar_region').push('not_a_ds')
+        assert next(victim_gen) == expected
```

Comment:
Since you're providing 3 fake years of data, I'd much prefer if your test didn't reinitialize the calculation every time. Several applications have memory (so this would be useful for them), and we should make sure that calculations that aren't supposed to have memory work as expected.

Reply:
Resolved this with 3a56bff6af363e2b4f2735b82abbb422b682b2a5

The test mentioned tests for correct behavior outside, within, and along the edge of the given clip interval.

I added a new test that explicitly tests memory. We give three years of input in a single test instance and ensure that each iteration of the generator gives the correct output. I believe this should deal with this case.